packages/react-native-executorch/common/rnexecutorch/models/voice_activity_detection/VoiceActivityDetection.cpp

```diff
@@ -0,0 +1,159 @@
+#include "VoiceActivityDetection.h"
+#include "rnexecutorch/models/voice_activity_detection/Utils.h"
+
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <numeric>
+#include <ranges>
+#include <vector>
+
+namespace rnexecutorch::models::voice_activity_detection {
+using namespace constants;
+using executorch::aten::Tensor;
+using executorch::extension::TensorPtr;
+
+VoiceActivityDetection::VoiceActivityDetection(
+    const std::string &modelSource,
+    std::shared_ptr<react::CallInvoker> callInvoker)
+    : BaseModel(modelSource, callInvoker) {}
+
+std::vector<std::array<float, kPaddedWindowSize>>
+VoiceActivityDetection::preprocess(std::span<float> waveform) const {
+  auto kHammingWindowArray = utils::generateHammingWindow();
+
+  const size_t numFrames = (waveform.size() - kWindowSize) / kHopLength;
+
+  std::vector<std::array<float, kPaddedWindowSize>> frameBuffer(
+      numFrames, std::array<float, kPaddedWindowSize>{});
+
+  constexpr size_t totalPadding = kPaddedWindowSize - kWindowSize;
+  constexpr size_t leftPadding = totalPadding / 2;
+  constexpr size_t rightPadding = totalPadding - leftPadding;
+  for (size_t i = 0; i < numFrames; i++) {
+
+    auto windowView = waveform.subspan(i * kHopLength, kWindowSize);
+    std::ranges::copy(windowView, frameBuffer[i].begin() + leftPadding);
```

Comment:
Use `namespace ranges = std::ranges;` at the top of the file and the use ranges::copy etc. elsewhere